File: frontend/src/stores/project.ts
/**
 * Copyright (c) 2024 Tempo Budget
 * SPDX-License-Identifier: MIT
 *
 * Project Store
 *
 * Pinia store for managing project budget state and operations.
 */

import { defineStore } from 'pinia'
import { ref } from 'vue'
import {
  projectsAPI,
  type Project, type ProjectSummary, type ProjectPlannedExpense,
  type ProjectReminder, type ProjectTransaction, type ProjectCategoryWithSpent,
  type ProjectMemberWithUser, type ProjectInvitationWithDetails,
  type ProjectMemberBalance,
} from '@/services/api'

export const useProjectStore = defineStore('project', () => {
  const projects = ref<Project[]>([])
  const summaries = ref<ProjectSummary[]>([])
  const currentProject = ref<Project | null>(null)
  const plannedExpenses = ref<ProjectPlannedExpense[]>([])
  const projectTransactions = ref<ProjectTransaction[]>([])
  const reminders = ref<ProjectReminder[]>([])
  const loading = ref(false)

  // ── Projects ──

  async function fetchProjects(params?: { status?: string; mode?: string }) {
    loading.value = true
    try {
      projects.value = await projectsAPI.getAll(params)
    } finally {
      loading.value = false
    }
  }

  async function fetchSummaries() {
    loading.value = true
    try {
      summaries.value = await projectsAPI.getSummaries()
    } finally {
      loading.value = false
    }
  }

  async function fetchProject(id: string) {
    loading.value = true
    try {
      currentProject.value = await projectsAPI.getById(id)
    } finally {
      loading.value = false
    }
  }

  async function createProject(data: { name: string; description?: string; target_date?: string; total_budget: number; mode: 'personal' | 'pro' }) {
    const project = await projectsAPI.create(data)
    projects.value.unshift(project)
    return project
  }

  async function updateProject(id: string, data: { name?: string; description?: string; target_date?: string; total_budget?: number; status?: string }) {
    const updated = await projectsAPI.update(id, data)
    const idx = projects.value.findIndex(p => p.id === id)
    if (idx !== -1) projects.value[idx] = updated
    if (currentProject.value?.id === id) currentProject.value = updated
    return updated
  }

  async function deleteProject(id: string) {
    await projectsAPI.delete(id)
    projects.value = projects.value.filter(p => p.id !== id)
    if (currentProject.value?.id === id) currentProject.value = null
  }

  // ── Categories ──

  async function createCategory(projectId: string, data: { name: string; planned_amount: number }) {
    const category = await projectsAPI.createCategory(projectId, data)
    const newCat: ProjectCategoryWithSpent = {
      id: category.id,
      project_id: category.project_id,
      name: category.name,
      planned_amount: category.planned_amount,
      created_at: category.created_at,
      total_spent: 0,
      remaining: category.planned_amount,
    }

    // Update in currentProject
    if (currentProject.value?.id === projectId) {
      currentProject.value.categories.push(newCat)
    }

    // Also update in projects list
    const projectInList = projects.value.find(p => p.id === projectId)
    if (projectInList) {
      projectInList.categories.push({ ...newCat })
    }

    return category
  }

  async function updateCategory(projectId: string, categoryId: string, data: { name?: string; planned_amount?: number }) {
    const updated = await projectsAPI.updateCategory(projectId, categoryId, data)

    // Helper to update a category in a categories array
    const updateInArray = (categories: ProjectCategoryWithSpent[]) => {
      const idx = categories.findIndex(c => c.id === categoryId)
      if (idx !== -1) {
        const existing = categories[idx]!
        categories[idx] = {
          id: existing.id,
          project_id: existing.project_id,
          name: updated.name,
          planned_amount: updated.planned_amount,
          created_at: existing.created_at,
          total_spent: existing.total_spent,
          remaining: updated.planned_amount - existing.total_spent,
        }
      }
    }

    // Update in currentProject
    if (currentProject.value?.id === projectId) {
      updateInArray(currentProject.value.categories)
    }

    // Also update in projects list (used by HistoryView for category options)
    const projectInList = projects.value.find(p => p.id === projectId)
    if (projectInList) {
      updateInArray(projectInList.categories)
    }

    return updated
  }

  async function deleteCategory(projectId: string, categoryId: string) {
    await projectsAPI.deleteCategory(projectId, categoryId)

    // Update in currentProject
    if (currentProject.value?.id === projectId) {
      currentProject.value.categories = currentProject.value.categories.filter(c => c.id !== categoryId)
    }

    // Also update in projects list
    const projectInList = projects.value.find(p => p.id === projectId)
    if (projectInList) {
      projectInList.categories = projectInList.categories.filter(c => c.id !== categoryId)
    }
  }

  // ── Planned Expenses ──

  async function fetchPlannedExpenses(projectId: string) {
    plannedExpenses.value = await projectsAPI.getPlannedExpenses(projectId)
  }

  async function createPlannedExpense(projectId: string, data: { project_category_id: string; description: string; amount: number; due_date?: string; reminder_date?: string }) {
    const expense = await projectsAPI.createPlannedExpense(projectId, data)
    plannedExpenses.value.push(expense)
    return expense
  }

  async function updatePlannedExpense(projectId: string, expenseId: string, data: {
    project_category_id?: string; description?: string; amount?: number;
    due_date?: string; reminder_date?: string; status?: string;
    transaction_id?: string; pro_transaction_id?: string
  }) {
    const updated = await projectsAPI.updatePlannedExpense(projectId, expenseId, data)
    const idx = plannedExpenses.value.findIndex(e => e.id === expenseId)
    if (idx !== -1) plannedExpenses.value[idx] = updated
    return updated
  }

  async function deletePlannedExpense(projectId: string, expenseId: string) {
    await projectsAPI.deletePlannedExpense(projectId, expenseId)
    plannedExpenses.value = plannedExpenses.value.filter(e => e.id !== expenseId)
  }

  // ── Transactions ──

  async function fetchProjectTransactions(projectId: string) {
    projectTransactions.value = await projectsAPI.getTransactions(projectId)
  }

  // ── Reminders ──

  async function fetchReminders() {
    reminders.value = await projectsAPI.getReminders()
  }

  // ── Members ──

  const members = ref<ProjectMemberWithUser[]>([])
  const memberBalances = ref<ProjectMemberBalance[]>([])
  const projectInvitations = ref<ProjectInvitationWithDetails[]>([])

  async function fetchMembers(projectId: string) {
    members.value = await projectsAPI.getMembers(projectId)
  }

  async function inviteMember(projectId: string, email: string, role: string = 'member') {
    await projectsAPI.inviteMember(projectId, email, role)
  }

  async function removeMember(projectId: string, memberId: string) {
    await projectsAPI.removeMember(projectId, memberId)
    members.value = members.value.filter(m => m.id !== memberId)
  }

  async function updateMemberShare(projectId: string, memberId: string, share: number) {
    const updated = await projectsAPI.updateMemberShare(projectId, memberId, share)
    const idx = members.value.findIndex(m => m.id === memberId)
    if (idx !== -1) members.value[idx] = updated
    return updated
  }

  async function fetchMemberBalances(projectId: string) {
    memberBalances.value = await projectsAPI.getBalances(projectId)
  }

  async function fetchProjectInvitations() {
    projectInvitations.value = await projectsAPI.getMyInvitations()
  }

  async function acceptProjectInvitation(invitationId: string) {
    await projectsAPI.acceptInvitation(invitationId)
    projectInvitations.value = projectInvitations.value.filter(i => i.id !== invitationId)
  }

  async function rejectProjectInvitation(invitationId: string) {
    await projectsAPI.rejectInvitation(invitationId)
    projectInvitations.value = projectInvitations.value.filter(i => i.id !== invitationId)
  }

  return {
    projects, summaries, currentProject, plannedExpenses, projectTransactions, reminders, loading,
    members, memberBalances, projectInvitations,
    fetchProjects, fetchSummaries, fetchProject,
    createProject, updateProject, deleteProject,
    createCategory, updateCategory, deleteCategory,
    fetchPlannedExpenses, createPlannedExpense, updatePlannedExpense, deletePlannedExpense,
    fetchProjectTransactions, fetchReminders,
    fetchMembers, inviteMember, removeMember, updateMemberShare, fetchMemberBalances,
    fetchProjectInvitations, acceptProjectInvitation, rejectProjectInvitation,
  }
})

File: backend/app/models/project.py
# Copyright (c) 2024 Tempo Budget
# SPDX-License-Identifier: MIT
"""Project-related data models."""

from typing import Literal
from pydantic import BaseModel, Field


class ProjectCategory(BaseModel):
    """A category within a project with a planned budget amount."""
    id: str = Field(..., description="Unique identifier")
    project_id: str = Field(..., description="Project ID")
    name: str = Field(..., description="Category name")
    planned_amount: float = Field(..., description="Planned budget amount")
    created_at: str = Field(..., description="Creation timestamp")

    class Config:
        from_attributes = True


class ProjectCategoryWithSpent(ProjectCategory):
    """Project category with computed spent totals."""
    total_spent: float = Field(0, description="Total spent in this category")
    remaining: float = Field(0, description="Remaining budget")


class Project(BaseModel):
    """A budget project (e.g. wedding, renovation)."""
    id: str = Field(..., description="Unique identifier")
    user_id: str = Field(..., description="Owner user ID")
    name: str = Field(..., description="Project name")
    description: str | None = Field(None, description="Project description")
    target_date: str | None = Field(None, description="Target completion date")
    total_budget: float = Field(..., description="Total planned budget")
    status: str = Field(..., description="active, completed, or abandoned")
    mode: str = Field(..., description="personal or pro")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")
    total_spent: float = Field(0, description="Total spent across all categories")
    remaining: float = Field(0, description="Remaining budget")
    categories: list[ProjectCategoryWithSpent] = Field(default_factory=list, description="Project categories with spent")

    class Config:
        from_attributes = True


class CreateProject(BaseModel):
    """Request payload for creating a project."""
    name: str = Field(..., min_length=1, description="Project name")
    description: str | None = Field(None, description="Project description")
    target_date: str | None = Field(None, description="Target completion date")
    total_budget: float = Field(..., ge=0, description="Total planned budget")
    mode: Literal["personal", "pro"] = Field(..., description="Project mode")


class UpdateProject(BaseModel):
    """Request payload for updating a project."""
    name: str | None = Field(None, min_length=1, description="Project name")
    description: str | None = Field(None, description="Project description")
    target_date: str | None = Field(None, description="Target completion date")
    total_budget: float | None = Field(None, ge=0, description="Total planned budget")
    status: Literal["active", "completed", "abandoned"] | None = Field(None, description="Project status")


class CreateProjectCategory(BaseModel):
    """Request payload for creating a project category."""
    name: str = Field(..., min_length=1, description="Category name")
    planned_amount: float = Field(..., ge=0, description="Planned budget amount")


class UpdateProjectCategory(BaseModel):
    """Request payload for updating a project category."""
    name: str | None = Field(None, min_length=1, description="Category name")
    planned_amount: float | None = Field(None, ge=0, description="Planned budget amount")


class ProjectPlannedExpense(BaseModel):
    """A planned/future expense within a project."""
    id: str = Field(..., description="Unique identifier")
    project_id: str = Field(..., description="Project ID")
    project_category_id: str = Field(..., description="Project category ID")
    description: str = Field(..., description="Expense description")
    amount: float = Field(..., description="Planned amount")
    due_date: str | None = Field(None, description="Due date")
    reminder_date: str | None = Field(None, description="Reminder date")
    status: str = Field(..., description="pending or paid")
    transaction_id: str | None = Field(None, description="Linked personal transaction ID")
    pro_transaction_id: str | None = Field(None, description="Linked pro transaction ID")
    created_at: str = Field(..., description="Creation timestamp")
    updated_at: str = Field(..., description="Last update timestamp")
    category_name: str | None = Field(None, description="Resolved category name")

    class Config:
        from_attributes = True


class CreatePlannedExpense(BaseModel):
    """Request payload for creating a planned expense."""
    project_category_id: str = Field(..., description="Project category ID")
    description: str = Field(..., min_length=1, description="Expense description")
    amount: float = Field(..., gt=0, description="Planned amount")
    due_date: str | None = Field(None, description="Due date")
    reminder_date: str | None = Field(None, description="Reminder date")


class UpdatePlannedExpense(BaseModel):
    """Request payload for updating a planned expense."""
    project_category_id: str | None = Field(None, description="Project category ID")
    description: str | None = Field(None, min_length=1, description="Expense description")
    amount: float | None = Field(None, gt=0, description="Planned amount")
    due_date: str | None = Field(None, description="Due date")
    reminder_date: str | None = Field(None, description="Reminder date")
    status: Literal["pending", "paid"] | None = Field(None, description="Expense status")
    transaction_id: str | None = Field(None, description="Linked personal transaction ID")
    pro_transaction_id: str | None = Field(None, description="Linked pro transaction ID")


class ProjectSummary(BaseModel):
    """Summary of a project for list views."""
    id: str = Field(..., description="Project ID")
    name: str = Field(..., description="Project name")
    mode: str = Field(..., description="personal or pro")
    status: str = Field(..., description="Project status")
    total_budget: float = Field(..., description="Total planned budget")
    total_spent: float = Field(0, description="Total spent")
    remaining: float = Field(0, description="Remaining budget")
    percentage: float = Field(0, description="Percentage spent")
    category_count: int = Field(0, description="Number of categories")
    planned_expense_count: int = Field(0, description="Total planned expenses")
    pending_expense_count: int = Field(0, description="Pending planned expenses")
    target_date: str | None = Field(None, description="Target completion date")


class ProjectReminder(BaseModel):
    """A reminder for an upcoming planned expense."""
    id: str = Field(..., description="Planned expense ID")
    project_id: str = Field(..., description="Project ID")
    project_name: str = Field(..., description="Project name")
    description: str = Field(..., description="Expense description")
    amount: float = Field(..., description="Planned amount")
    due_date: str | None = Field(None, description="Due date")
    reminder_date: str = Field(..., description="Reminder date")
    category_name: str | None = Field(None, description="Category name")
